core/src/main/java/org/apache/iceberg/OverwriteData.java

```diff
@@ -88,6 +121,40 @@ public OverwriteFiles validateAddedFiles() {
       }
     }
 
+    if (conflictingAppendsRowFilter != null) {
+      PartitionSpec spec = writeSpec();
+      Expression inclusiveExpr = Projections.inclusive(spec).project(conflictingAppendsRowFilter);
+      Evaluator inclusive = new Evaluator(spec.partitionType(), inclusiveExpr);
+
+      List<DataFile> newFiles = collectNewFiles(base);
+      for (DataFile newFile : newFiles) {
+        // we do partition-level conflict resolution right now
+        // we can enhance it by leveraging column stats and MetricsEvaluator
+        ValidationException.check(
+            !inclusive.eval(newFile.partition()),
+            "A conflicting file was appended that matches filter '%s': %s",
+            conflictingAppendsRowFilter, newFile.path());
+      }
+    }
+
     return super.apply(base);
   }
+
+  private List<DataFile> collectNewFiles(TableMetadata meta) {
+    List<DataFile> newFiles = new ArrayList<>();
+
+    Long currentSnapshotId = meta.currentSnapshot() == null ? null : meta.currentSnapshot().snapshotId();
+    while (currentSnapshotId != null && !currentSnapshotId.equals(readSnapshotId)) {
+      Snapshot currentSnapshot = meta.snapshot(currentSnapshotId);
+
+      if (currentSnapshot == null) {
+        throw new ValidationException("Cannot find snapshot %d. Was it expired?", currentSnapshotId);
```

Comment:
If `meta.currentSnapshot` is null, this will print "Cannot find snapshot null". I think it would be better to start the method with a precondition that checks that meta has a current snapshot and fails with a better error message if it doesn't.

Also, there are a few reasons why the snapshot can't be found. The base snapshot may have been expired, but it also could have been rolled back. I think this should just state "Cannot determine history between base snapshot %s and current %s".

Reply:
I think `meta.currentSnapshot` can be `null` if the overwrite happens on an empty table. In that case, the read snapshot is also `null`, so it should work correctly. We have a test for this as well. Such a situation might happen if we use this API for eager `MERGE INTO` operations.

Will update the error message.

Reply:
I agree that this would work correctly, but "Cannot find snapshot null" isn't a very helpful error message. I think it makes sense to catch that case earlier and have a better error.